zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,634 @@
+    pub(crate) type OrderOf<T> = Order<AccountIdOf<T>, BalanceOf<T>, MarketIdOf<T>>;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+    pub(crate) type MomentOf<T> = <<T as Config>::MarketCommons as MarketCommonsPalletApi>::Moment;
+    pub(crate) type MarketOf<T> =
+        Market<AccountIdOf<T>, BalanceOf<T>, BlockNumberFor<T>, MomentOf<T>, Asset<MarketIdOf<T>>>;
+    pub(crate) type OrdersOf<T> = BoundedVec<OrderId, <T as Config>::MaxOrders>;
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T>
+    where
+        T: Config,
+    {
+        /// A trade was executed.
+        HybridRouterExecuted {
+            tx_type: TxType,
+            who: AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            price_limit: BalanceOf<T>,
+            asset_in: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            asset_out: AssetOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The specified amount is zero.
+        AmountIsZero,
+        /// The price limit is too high.
+        PriceLimitTooHigh,
+        /// The price of an order is above the specified maximum price.
+        OrderPriceAboveMaxPrice,
+        /// The price of an order is below the specified minimum price.
+        OrderPriceBelowMinPrice,
+        /// The asset of an order is not equal to the maker asset of the order book.
+        AssetNotEqualToOrderBookMakerAsset,
+        /// The asset of an order is not equal to the taker asset of the order book.
+        AssetNotEqualToOrderBookTakerAsset,
+        /// The strategy "immediate or cancel" was applied.
+        CancelStrategyApplied,
+        /// The asset count does not match the markets asset count.
+        AssetCountMismatch,
+        /// The maximum number of orders was exceeded.
+        MaxOrdersExceeded,
+        /// Action cannot be completed because an unexpected error has occurred. This should be
+        /// reported to protocol maintainers.
+        InconsistentState(InconsistentStateError),
+    }
+
+    // NOTE: these errors should never happen.
+    #[derive(Encode, Decode, Eq, PartialEq, TypeInfo, frame_support::PalletError, RuntimeDebug)]
+    pub enum InconsistentStateError {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Routes a buy order to AMM and CDA to achieve the best average execution price.
+        ///
+        /// # Parameters
+        ///
+        /// * `market_id`: The ID of the market to buy from.
+        /// * `asset_count`: The number of assets traded on the market.
+        /// * `asset`: The asset to buy.
+        /// * `amount_in`: The amount of the market's base asset to sell.
+        /// * `max_price`: The maximum price to buy at.
```

Comment:
Can you suggest something?

Reply:
"The maximum allowed value to move the price to by this trade." Not sure if that's clearer though. Feel free to use either one.

Reply:
But this description only holds true for the AMM. You can't move the price of an order in the order book. The maximum price to buy at seems correct to me, since this is exactly what it is. It is a limit of how much you are willing to pay for the outcome asset.

Reply:
That's true, in case of the order book, you'd be moving the smallest ask or largest bid. Let's just keep it this way.